lfs/config.go

```diff
@@ -151,20 +162,94 @@ func NewEndpointFromCloneURL(url string) Endpoint {
 }
 
 // Create a new endpoint from a general URL
-func NewEndpoint(url string) Endpoint {
-	e := Endpoint{Url: url}
-
-	if !httpPrefixRe.MatchString(url) {
-		pieces := strings.SplitN(url, ":", 2)
-		hostPieces := strings.SplitN(pieces[0], "@", 2)
-		if len(hostPieces) == 2 {
-			e.SshUserAndHost = pieces[0]
-			e.SshPath = pieces[1]
-			e.Url = fmt.Sprintf("https://%s/%s", hostPieces[1], pieces[1])
+func NewEndpoint(rawurl string) Endpoint {
+	var endpoint Endpoint
+	// Check for SSH URLs
+	// Support ssh://user@host.com/path/to/repo and user@host.com:path/to/repo
+	u, err := url.Parse(rawurl)
+	if err == nil {
+		u = processBareSshUrl(u)
+		switch u.Scheme {
+		case "ssh":
+			endpoint = endpointFromSshUrl(u)
+		case "http", "https":
+			endpoint = endpointFromHttpUrl(u)
+		default:
+			// Just passthrough to preserve
+			endpoint = Endpoint{Url: rawurl}
 		}
 	}
 
-	return e
+	return endpoint
+}
+
+// For ease of processing, sanitise a bare Git SSH URL into a ssh:// URL
+func processBareSshUrl(u *url.URL) *url.URL {
+	if u.Scheme == "" && u.Path != "" {
+		// This might be a bare SSH URL
+		// Turn it into ssh:// for simplicity of extraction later
+		parts := strings.Split(u.Path, ":")
+		if len(parts) > 1 {
+			// Treat presence of ':' as a bare URL
+			var newPath string
+			if len(parts) > 2 { // port included; really should only ever be 3 parts
+				newPath = fmt.Sprintf("%v:%v", parts[0], strings.Join(parts[1:], "/"))
+			} else {
+				newPath = strings.Join(parts, "/")
+			}
+			newrawurl := fmt.Sprintf("ssh://%v", newPath)
+			newu, err := url.Parse(newrawurl)
+			if err == nil {
+				return newu
+			}
+		}
+	}
+	return u
```

Comment:
I'd prefer early returns.
```go
if u.Scheme != "" || u.Path == "" {
	return u
}

// This might be a bare SSH URL
// Turn it into ssh:// for simplicity of extraction later
parts := strings.Split(u.Path, ":")
if len(parts) < 2 {
	return u
}

// Treat presence of ':' as a bare URL
var newPath string
if len(parts) > 2 { // port included; really should only ever be 3 parts
	newPath = fmt.Sprintf("%v:%v", parts[0], strings.Join(parts[1:], "/"))
} else {
	newPath = strings.Join(parts, "/")
}

newrawurl := fmt.Sprintf("ssh://%v", newPath)
newu, err := url.Parse(newrawurl)
if err != nil {
	return u
}

return newu
```